tribits/doc/guides/TribitsGuidesBody.rst

```diff
@@ -6426,44 +6451,67 @@ There are cases where one may need to enable or disable some feature that
+examples such as support for a given data-type that may impact multiple
+downstream packages.
 
 When the internal configuration of a package (i.e. while processing its
 ``<packageDir>/CMakeLists.txt`` file) determines that an optional feature
-``<XXX>_ENABLE_<YYY>`` must be enabled or disabled with and will change the
-value previously set (e.g. during the "Adjust package and TPLs enables and
-disables" stage), one cannot use a simple ``set()`` statement.  Changing the
-value of an ``<XXX>_ENABLE_<YYY>`` variable inside a package's
-``<packageDir>/CMakeLists.txt`` file using a raw ``set(<XXX>_ENABLE_<YYY>
+``<Package>_ENABLE_<YYY>`` must change the value previously set (e.g. that was
+set automatically by TriBITS during the "Adjust package and TPLs enables and
+disables" stage in `Full Processing of TriBITS Project Files`_), one cannot
+use a simple ``set()`` statement.  Changing the value of a
+``<Package>_ENABLE_<YYY>`` variable inside a package's
+``<packageDir>/CMakeLists.txt`` file using a raw ``set(<Package>_ENABLE_<YYY>
 <newValue>)`` statement only changes the variable's value inside the package's
 scope, but all other packages will see the old value of
-``<XXX>_ENABLE_<YYY>``.  To correctly change the value of one of these
+``<Package>_ENABLE_<YYY>``.  To correctly change the value of one of these
 variables, instead use `dual_scope_set()`_ from the top-level
-``<packageDir>/CMakeLists.txt`` file.  This sets the value in both the
-base-level (global) project scope and in the local scope of
-``<packageDir>/CMakeLists.txt``.  (But this does **not** change the value of a
-cache variable ``<XXX>_ENABLE_<YYY>`` that may have been set by the user or
-some other means; see `TriBITS auto-enables/disables done using non-cache
-local variables`_.)  Any downstream package (configured after processing
-``<packageDir>/CMakeLists.txt``) will see the new value ``<XXX>_ENABLE_<YYY>
-STREQUAL <val>``.  It is also strongly recommended that a message or warning
-be printed to CMake STDOUT using ``message(["NOTE: "|WARNING] "<message>")``
-when globally changing an ENABLE variable. The user may have set it
-explicitly, and they should know exactly why and where their choice is being
-overridden.
+``<packageDir>/CMakeLists.txt`` file. And when changing the value of a
+``<Package>_ENABLE_<UpstreamPackage>`` variable, one must also change the
+value of the associated
+`HAVE_<PACKAGE_NAME_UC>_<OPTIONAL_DEP_PACKAGE_NAME_UC>`_ variable using
+``deal_scope_set()`` as well.  For example, to disable optional support for
+``<UpstreamPackage>`` in ``<DownstreamPackage>`` in ``<DownstreamPackage>``
+package's ``<packageDir>/CMakeLists.txt`` file, add the CMake code::
+
+  if (<some-condition>)
+    message("-- " "NOTE: ${PACKAGE_NAME}_ENABLE_<UpstreamPackage> being set to OFF"
+      " because of <reason>" )
+    dual_scope_set(${PACKAGE_NAME}_ENABLE_<UpstreamPackage> OFF)
+    dual_scope_set(HAVE_${PACKAGE_NAME_UC}_<UPSTREAM_PACKAGE_UC> OFF)
```

Comment:
Since this is messing with dependencies, and since one usually gets some nice output of what's going on during the initial parsing of dependencies: would it make sense to wrap this in a tiny utility function? Something like `tribits_disable_optional_dependency(UpstreamPackage, reason)`?

Reply:
> Something like `tribits_disable_optional_dependency(UpstreamPackage, reason)`?

@cgcgcg, that is a good idea.  One would think this is a rare situation but there must have been others or I would have never written that section in the documentation in the first place.  If it is worth writing the documentation, then it should be worth providing a macro do to this (and it would provide some layer of encapsulation as well).

It would take work to provide an example and testing for this new command.  But this command is small and simple enough and it would be used rarely enough that it may not be worth the effort to do so.

Reply:
You might be right. Grepping for `dual_scope_set` in Trilinos gives only two occurrences, one in ML and the one in Xpetra.

Reply:
Yup, that is all I can find:

```
$ find packages/ \( -name "*.cmake" -or -name "CMakeLists.txt" \) -exec grep -nHi dual_scope_set {} \;
packages/ml/CMakeLists.txt:33:  DUAL_SCOPE_SET(${PACKAGE_NAME}_ENABLE_Zoltan OFF)
packages/ml/CMakeLists.txt:34:  DUAL_SCOPE_SET(HAVE_ML_ZOLTAN OFF)
packages/xpetra/CMakeLists.txt:11:DUAL_SCOPE_SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${PACKAGE_SOURCE_DIR}/cmake)
packages/xpetra/CMakeLists.txt:93:  DUAL_SCOPE_SET(${PACKAGE_NAME}_ENABLE_EpetraExt OFF)
packages/xpetra/CMakeLists.txt:94:  DUAL_SCOPE_SET(HAVE_XPETRA_EPETRAEXT OFF)
```

Looking at the Git logs, I added those statements to `packages/ml/CMakeLists.txt` **14 years ago**.

I seem to remember some tweaks being done in SEACAS that motivated me to write the documentation, but I don't seem to be able to find this.

Reply:
I will try adding `tribits_disable_optional_dependency()` in the Trilinos copy of TriBITS and I can add some more checking like it must be called in the top-level base package `CMakeLists.txt` file and not in a subpackage `CMakeLists.txt` file (since PARENT_SCOPE will not be setting the base project-level scope).

Reply:
Are you sure that it's worth it for essentially one occurrence in Trilinos? (Assuming that ML will go away at some point.)

Reply:
FYI: `tribits_disable_optional_dependency()` is used in the updated doc in new commit [`220040d` (#566)](https://github.com/TriBITSPub/TriBITS/pull/566/commits/220040d6ed9d981095977d540bac89a90c1f8f70).  See the updated section [here](https://github.com/bartlettroscoe/TriBITS/blob/tril-11612-doc-have_package_upstream/tribits/doc/guides/TribitsGuidesBody.rst#how-to-tweak-downstream-tribits-enable-variables-during-package-configuration).